Observing Planning & Follow-up PR #3: Observing Planner

dannygoldstein opened the pull request:
![Jul-27-2020 22-51-24](https://user-images.githubusercontent.com/2769632/88624430-ce4d5f80-d05b-11ea-8985-968277b6c086.gif)

This PR introduces the Observing Planner frontend to skyportal. The observing planner (OP) is the go-to tool for observers at the telescope observing Skyportal targets. The OP renders all of the assignments for a given observing run in a searchable, filterable table with links to finder charts and tabulated ephemeris data. The table is responsive, renders well on mobile, and provides an interactive interface for observers to mark their targets observed / not observed, and upload the resulting data. 

The OP displays cutouts, and javascript airmass chart, and a lightweight JS lightcurve for each object. 

The table enforces ACLs and only allows observers to see objects that they have permission to view. The OP makes use of the infrastructure and sky calculations added in #591 , #574 , and #559. 

The table currently re-renders specific sources in response to the REFRESH_SOURCE action, but in a bit of a hacky way. In the future a cleanup of the state tree structure should provide a more seamless way to share data between pages.

dannygoldstein commented:
Unable to reproduce failures on travis locally. I do not believe they are not related to this PR.

dannygoldstein commented:
[ch527]

dannygoldstein commented:
Locally all tests pass for this PR: 

`===================== 157 passed, 1 xpassed, 32 warnings in 505.31s (0:08:25) =====================`

profjsb commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,244 @@
+        try:
+            assignment = ClassicalAssignment(**AssignmentSchema.load(data=data))
+        except ValidationError as e:
+            return self.error(f'Error parsing followup request: '
```
```suggestion
            return self.error('Error parsing followup request: '
```

profjsb commented on skyportal/handlers/api/observingrun.py:
```diff
@@ -0,0 +1,232 @@
+        DBSession().commit()
+        return self.success()
+
+    @permissions(['Upload data'])
```
Deleting data has a very different meaning that uploading data. Should we have a new ACL = 'Delete data'?

dannygoldstein replied:
Sounds reasonable to me

dannygoldstein replied:
Opened a new issue for this

profjsb commented on static/js/components/AssignmentForm.jsx:
```diff
@@ -0,0 +1,155 @@
+  const groups = useSelector((state) => state.groups.all);
+
+  const upcomingRuns = observingRunList.filter((observingrun) => (
+    Date.parse(observingrun.ephemeris.sunrise_utc) >= Date.now()
```
Do we need more complex logic to account for timezones? I think this might get a different behavior if run simultaneously in Europe and the US while it's nighttime in Hawaii. Perhaps use a Date.now() minus 12 hours to be safe, to capture ongoing runs?

dannygoldstein replied:
`Date.now()` returns the UNIX timestamp in ms, which is natively UTC.

profjsb commented on static/js/components/AssignmentForm.jsx:
```diff
@@ -0,0 +1,155 @@
+              rules={{ required: true }}
+            >
+              {
+                ["1", "2", "3", "4", "5"].map((prio) => (
```
Does this need to be hardcoded? Or should the set of possibilities be grabbed from the state? I can imagine a different SP instance using different priority nomenclature.

dannygoldstein replied:
Deferred to #603

profjsb commented on static/js/components/ThumbnailList.jsx:
```diff
@@ -3,31 +3,71 @@ import PropTypes from 'prop-types';
+/* const useStyles = makeStyles((theme) => ({
+})); */
+
+const useStyles = makeStyles(() => ({
+  root: {
+    width: 200,
+    margin: 'auto',
+    maxHeight: 500,
+    flexGrow: 1
+  },
+  title: {
+    fontSize: 14,
+  },
+  pos: {
+    marginBottom: 0,
+  },
+  mediaDiv: {
+    position: "relative"
+  },
+  media: {
+    height: 200,
+    width: 200,
+  },
+  crosshair: {
+    position: "absolute",
+    top: 0,
+    left: 0,
+    width: 200,
+    height: 200,
+    paddingBottom: "0.2em"
+  }
+}));
```
Why not put this in `ThumnailList.css`?

dannygoldstein replied:
https://material-ui.com/styles/basics/#why-use-material-uis-styling-solution

profjsb commented on static/js/components/VegaPlot.jsx:
```diff
@@ -56,7 +56,8 @@ const spec = (url) => ({
           field: "filter",
-          type: "nominal"
+          type: "nominal",
+          scale: { domain: ["ztfg", "ztfr", "ztfi"], range: ["#377E22", "#EA3323", "#CCCC52"] }
```
I'm ok with an opinionated set of color assignments to different filters but shouldn't these all live in one place (e.g. the DB/enum file) and be grabbed dynamically from there? Also, what if the photometry is not from ztf (e.g. it's a Gatinni-only source)?

dannygoldstein replied:
Nice catch. This is an important thing to fix. We need to define the color that shows up for each filter, otherwise filters with undefined colors will not be plotted. Will push up a fix

dannygoldstein replied:
Just reverting to the Vega defaults for now (not specifying any specific colors for specific filters) which should fix the problem for the moment

profjsb commented on static/js/ducks/observingRun.js:
```diff
@@ -0,0 +1,68 @@
+
+  if (actionType === REFRESH_SOURCE) {
+    const { obj_id } = payload;
+    const assignment = observingRun.assignments.filter((a) => (a.obj_id === obj_id))[0];
```
if the filter fails, wont it leave behind an empty list, which then throws an error when indexed to 0?

dannygoldstein replied:
Nope, in JS indexing errors do not halt the program, they just return `undefined`. And this I check for on the next line.

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,428 @@
+    });
+
+
+    const renderObjId = (dataIndex) => {
```
Again, it'd be helpful to add comments clarifying that these are just for the MUI datatables `options`/`columns`

dannygoldstein replied:
Fixed

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,428 @@
+                  <em>Observing Planner for:</em>
+                </Typography>
+                <Typography variant="h4" gutterBottom color="textSecondary">
+                  <b>{observingRunTitle(observingRun, instrumentList, telescopeList, groups)}</b>
```
Make into a React component

dannygoldstein replied:
`observingRunTitle` just returns a string -- is it really necessary to make that a reactComponent?

acrellin replied:
Ah, no in that case it's fine. Disregard.

acrellin commented on static/js/components/AssignmentForm.jsx:
```diff
@@ -0,0 +1,155 @@
+
+import 'react-datepicker/dist/react-datepicker-cssmodules.css';
+
+export function observingRunTitle(observingRun, instrumentList, telescopeList, groups) {
```
Let's make this a React component that takes a `props` object

dannygoldstein replied:
See comment above

acrellin commented on static/js/components/AssignmentForm.jsx:
```diff
@@ -0,0 +1,155 @@
+  return `${observingRun?.calendar_date} ${instrument?.name}/${telescope?.nickname} (PI: ${observingRun?.pi} ${group?.name})`;
+}
+
+function makeMenuItem(observingRun, instrumentList, telescopeList, groups) {
```
Refactor to React component

acrellin replied:
Or simply move to the main return statement as it's so small

dannygoldstein replied:
Fixed

acrellin commented on static/js/components/StarList.jsx:
```diff
@@ -1,11 +1,29 @@
 
 import styles from './StarList.css';
 
+function starListElem(starList) {
```
Again, prefer these to be React components

acrellin replied:
Also, does `Elem` mean "element"? If so, it looks like the name is misleading, as it doesn't just render a single element of the starlist, but rather every element in the starlist. Or the name could just be more explicit/clearer

dannygoldstein replied:
Made this a react component called `StarListBody`.

acrellin commented on static/js/components/StarList.jsx:
```diff
@@ -21,25 +39,49 @@ const StarList = ({ sourceId }) => {
+        )
+      );
+
+      const starlist_info = [];
```
camelCase

dannygoldstein replied:
fixed

acrellin commented on static/js/components/StarList.jsx:
```diff
@@ -21,25 +39,49 @@ const StarList = ({ sourceId }) => {
+    };
+    fetchStarList();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [observingRunId, dispatch]);
```
`observingRunId` isn't even used in the hook, is it? shouldn't `assignments` be in the deps array?

dannygoldstein replied:
Yeah. I only want to render this once every page load, so I will reimplement this without using `useEffect`.

dannygoldstein replied:
Fixed

acrellin commented on static/js/ducks/observingRun.js:
```diff
@@ -0,0 +1,68 @@
+      const assignment_ids = state.assignments.map((a) => a.id);
+      if (assignment_ids.includes(assignment.id)) {
+        const index = assignment_ids.indexOf(assignment.id);
+        state.assignments[index] = assignment;
```
`state` should be immutable -- need to declare a new object and then operate on that, then `return { ...state, ...newState };`

dannygoldstein replied:
Fixed

acrellin commented on static/js/ducks/telescopes.js:
```diff
@@ -0,0 +1,26 @@
+const reducer = (state={ telescopeList: [] }, action) => {
+  switch (action.type) {
+    case FETCH_TELESCOPES_OK: {
+      const telescopes = action.data;
```
If you do `const telescopeList = action.data;` then below, can do shorthand object literal: `{ ...state, telescopeList }`

dannygoldstein replied:
fixed

acrellin commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,244 @@
+        out = ClassicalAssignment.__schema__().dump(screened, many=True)
+
+        # calculate when the targets rise and set
+        for d, a in zip(out, screened):
```
How about more descriptive variable names here, though, to make it clearer what's going on?

dannygoldstein replied:
Fixed

dannygoldstein commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,244 @@
+        return self.success(data={"id": assignment.id})
+
+    @auth_or_token
+    def put(self, assignment_id):
```
Generally for updates we use PUT

dannygoldstein commented on configs/db_demo.yaml:
```diff
@@ -63,21 +64,69 @@ telescopes:
+    lat: 19.8283
+    lon: -155.478
+    name: Keck II 10m
+    id: Keck2
```
Fixed in order to sync up with data loader branch

dannygoldstein commented on configs/telescopes.yaml:
```diff
@@ -5,423 +5,484 @@
   name: Hobby-Eberly Telescope
   id: HET
```
Fixed in order to sync up with data loader branch

dannygoldstein commented on skyportal/handlers/api/internal/plot.py:
```diff
@@ -30,3 +35,27 @@ def get(self, obj_id):
+            self.error('Invalid assignment id.')
+
+        sunset = assignment.run.sunset
+        sunrise = assignment.run.sunrise
```
Yes. `astropy.time.Time`

dannygoldstein commented on skyportal/handlers/api/observingrun.py:
```diff
@@ -0,0 +1,232 @@
+        return self.success(data=out)
+
+    @permissions(['Upload data'])
+    def put(self, run_id):
```
We should probably be using PATCH, but keeping PUT here for consistency with other endpoints

dannygoldstein commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,244 @@
+                joinedload(ClassicalAssignment.requester),
+                joinedload(ClassicalAssignment.obj)
+                .joinedload(Obj.comments)
+            ).limit(1)
```
@dmitryduev  Yes, only using limit here as it reduces the amount of code needed downstream to handle scalar vs list query outputs

dannygoldstein replied:
@acrellin Fixed

dannygoldstein commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,244 @@
+
+        for a in assignments:
+            obj = Source.get_obj_if_owned_by(a.obj.id, self.current_user)
+            if obj is not None or 'System admin' in acls:
```
Nice catch. Deleting this as it is no longer needed thanks to your implementation of `User.associated_groups`

dannygoldstein commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,244 @@
+        screened = []
+
+        for a in assignments:
+            obj = Source.get_obj_if_owned_by(a.obj.id, self.current_user)
```
Did this always throw an error? I thought it used to just return `None` in the cases where it now raises?

dannygoldstein replied:
Fixed

dannygoldstein commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,244 @@
+        DBSession().add(assignment)
+        DBSession().commit()
+        self.push_all(
+            action="skyportal/REFRESH_SOURCE",
```
Yes.

dannygoldstein commented on skyportal/handlers/api/observingrun.py:
```diff
@@ -0,0 +1,232 @@
+                d['set_time_utc'] = a.set_time.isot
+
+            return self.success(data=data)
+        runs = ObservingRun.query.all()
```
Everyone can view all runs (telescope schedules are public). They can't necessarily view all run assignments.

dannygoldstein commented on skyportal/handlers/api/observingrun.py:
```diff
@@ -0,0 +1,232 @@
+                              'the run.')
+
+        DBSession().query(ObservingRun).filter(
+            ObservingRun.id == int(run_id)
```
fixed.

dannygoldstein commented on skyportal/handlers/api/telescope.py:
```diff
@@ -66,32 +66,54 @@ def post(self):
+            else:
+                return self.error(
+                    f"Could not load telescope with ID {telescope_id}")
         else:
```
Fixed on master

dannygoldstein commented on skyportal/handlers/api/telescope.py:
```diff
@@ -66,32 +66,54 @@ def post(self):
+            t = Telescope.get_if_owned_by(int(telescope_id), self.current_user)
+            if t is not None:
+                return self.success(data=t)
+            else:
```
Fixed on master

dannygoldstein commented on skyportal/models.py:
```diff
@@ -748,4 +764,150 @@ class FollowupRequest(Base):
+    )
+
+    # name of the PI
+    pi = sa.Column(sa.Text)
```
No, fixed.

dannygoldstein commented on static/js/components/AssignmentForm.jsx:
```diff
@@ -0,0 +1,155 @@
+
+  const group = groups?.filter((g) => g.id === observingRun.group_id)[0];
+
+  return `${observingRun?.calendar_date} ${instrument?.name}/${telescope?.nickname} (PI: ${observingRun?.pi} ${group?.name})`;
```
Agreed. Thanks.

dannygoldstein commented on skyportal/tests/frontend/test_assignments.py:
```diff
@@ -0,0 +1,29 @@
+        '//*[@name="assignmentSubmitButton"]'
+    )
+
+    driver.scroll_to_element_and_click(submit_button)
```
Fixed.

dannygoldstein commented on static/js/components/AssignmentForm.jsx:
```diff
@@ -0,0 +1,155 @@
+    const formData = {
+      // Need to add obj_id, etc to form data for request
+      ...initialFormState,
+      ...getValues({ nest: true })
```
Fixed.

dannygoldstein commented on static/js/components/AssignmentList.jsx:
```diff
@@ -0,0 +1,160 @@
+        </a>
+      </td>
+      <td>
+        {requester?.username || load}
```
I'd rather have "Loading..." render than just blank cells.

dannygoldstein commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,428 @@
+                useGrid={false}
+              />
+              <Grid item>
+                <Suspense fallback={<div>Loading plot...</div>}>
```
Just following what is done in `VegaPlot.jsx`.

dannygoldstein commented on static/js/components/AssignmentList.jsx:
```diff
@@ -0,0 +1,160 @@
+import styles from './AssignmentList.css';
+
+
+function renderAssignment(assignment, deleteAssignment, dispatch, users, observingRunList,
```
Fixed

dannygoldstein commented on static/js/components/AssignmentList.jsx:
```diff
@@ -0,0 +1,160 @@
+        <tbody>
+          {
+            assignments.map(
+              (assignment) => renderAssignment(
```
Fixed

dannygoldstein commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,428 @@
+  const ispending = assignment.status === "pending";
+  const isdone = assignment.status === "complete";
+
+  const complete_action = () => {
```
Fixed.

dannygoldstein commented on skyportal/handlers/api/observingrun.py:
```diff
@@ -0,0 +1,232 @@
+                return self.error(f"Could not load observing run {run_id}",
+                                  data={"run_id": run_id})
+            # order the assignments by ra
+            run.assignments = sorted(run.assignments, key=lambda a: a.obj.ra)
```
This is already a float (this returns the PSQL db value (decimal degrees, float), not a `astropy.coordinates.SkyCoord`).

dannygoldstein commented:
This should be green except for some frontend flakiness with

```
FAILED skyportal/tests/frontend/test_groups.py::test_add_new_group_explicit_self_admin
```
@acrellin merging #677 should make this green. Aside from that, this is ready for review

dannygoldstein commented:
```
=================================== FAILURES ===================================
____________________ test_add_new_group_explicit_self_admin ____________________
driver = <baselayer.app.test_util.MyCustomWebDriver (session="879045bd-536c-45ee-bd7c-207cb1fffa22")>
super_admin_user = <User(id=124, created_at=2020-08-09 00:11:07.837847, modified=2020-08-09 00:11:07.837847, username=dd1e582d-f0a8-478b-...2c9d717cf0af@cesium-ml.org, first_name=None, last_name=None, contact_email=None, contact_phone=None, preferences=None)>
user = <User(id=125, created_at=2020-08-09 00:11:07.865875, modified=2020-08-09 00:11:07.865875, username=b3318e2b-fa00-4a1d-...fc8507ed67e2@cesium-ml.org, first_name=None, last_name=None, contact_email=None, contact_phone=None, preferences=None)>
    def test_add_new_group_explicit_self_admin(driver, super_admin_user, user):
        test_proj_name = str(uuid.uuid4())
        driver.get(f'/become_user/{super_admin_user.id}')  # TODO decorator/context manager?
        driver.get('/')
        driver.refresh()
        driver.get('/groups')
        driver.wait_for_xpath('//input[@name="name"]').send_keys(test_proj_name)
        driver.wait_for_xpath('//input[@name="groupAdmins"]').send_keys(super_admin_user.username)
        driver.save_screenshot('/tmp/screenshot1.png')
        driver.wait_for_xpath('//input[@value="Create Group"]').click()
>       driver.wait_for_xpath(f'//a[contains(.,"{test_proj_name}")]')
skyportal/tests/frontend/test_groups.py:48: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
    expected_conditions.presence_of_element_located((By.XPATH, xpath)))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <selenium.webdriver.support.wait.WebDriverWait (session="879045bd-536c-45ee-bd7c-207cb1fffa22")>
method = <selenium.webdriver.support.expected_conditions.presence_of_element_located object at 0x7fc1b8554710>
message = ''
    def until(self, method, message=''):
        """Calls the method provided with the driver as an argument until the \
        return value is not False."""
        screen = None
        stacktrace = None
    
        end_time = time.time() + self._timeout
        while True:
            try:
                value = method(self._driver)
                if value:
                    return value
            except self._ignored_exceptions as exc:
                screen = getattr(exc, 'screen', None)
                stacktrace = getattr(exc, 'stacktrace', None)
            time.sleep(self._poll)
            if time.time() > end_time:
                break
>       raise TimeoutException(message, screen, stacktrace)
E       selenium.common.exceptions.TimeoutException: Message:
../../../virtualenv/python3.7.6/lib/python3.7/site-packages/selenium/webdriver/support/wait.py:80: TimeoutException
=========================== short test summary info ============================
FAILED skyportal/tests/frontend/test_groups.py::test_add_new_group_explicit_self_admin
============= 1 failed, 167 passed, 2 xpassed in 755.68s (0:12:35) =============
```

dannygoldstein commented:
Restarting the build...

dannygoldstein commented:
@stefanv @acrellin green and ready for review

acrellin commented:
@dannygoldstein looks like many of my initial review comments still apply and haven't been addressed. Will take another look when those are addressed.

dannygoldstein commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,428 @@
+
+  const render_items = [];
+  if (ispending) {
+    render_items.push(mark_observed);
```
Fixed

dannygoldstein commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,428 @@
+    </MenuItem>
+  );
+
+  const upload_photometry = (
```
Fixed

dannygoldstein commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,428 @@
+  } else {
+    const { assignments } = observingRun;
+
+    const renderPullOutRow = ((rowData, rowMeta) => {
```
Fixed

dannygoldstein commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,244 @@
+        source = Source.get_obj_if_owned_by(assignment.obj_id, self.current_user)
+        obj = Obj.query.get(assignment.obj_id)
+        acls = [a.id for a in self.current_user.acls]
+        if (obj is None) or (source is None and 'System admin' not in acls):
```
Fixed

dannygoldstein commented:
@acrellin @stefanv @profjsb Ready for re review !

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+import React, { useEffect, Suspense } from 'react';
+import PropTypes from 'prop-types';
+import { useSelector, useDispatch } from 'react-redux';
+
```
```suggestion
```

dannygoldstein replied:
Fixed

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+  const isdone = assignment.status === "complete";
+  const isnotdone = assignment.status === "not observed";
+
+  const completeAction = () => {
+    handleClose();
+    return dispatch(
+      SourceAction.editAssignment({ status: "complete" }, assignment.id)
```
```suggestion
  const updateAssignmentStatus = (status) => {
    handleClose();
    return dispatch(
      SourceAction.editAssignment({ status }, assignment.id)
```

dannygoldstein replied:
Done

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+    );
+  };
+
+  const pendingAction = () => {
+    handleClose();
+    return dispatch(
+      SourceAction.editAssignment({ status: "pending" }, assignment.id)
+    );
+  };
+
+  const notObservedAction = () => {
+    handleClose();
+    return dispatch(
+      SourceAction.editAssignment({ status: "not observed" }, assignment.id)
+    );
+  };
+
```
```suggestion
```

dannygoldstein replied:
Done

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+      >
+        {
+          (ispending || isnotdone) && (
+          <MenuItem onClick={completeAction} variant="contained" key={`${assignment.id}_done`}>
```
```suggestion
          <MenuItem
            onClick={() => updateAssignmentStatus("complete")}
            variant="contained"
            key={`${assignment.id}_done`}
          >
```

dannygoldstein replied:
Fixed

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+}
+        {
+          (ispending || isdone) && (
+          <MenuItem onClick={notObservedAction} variant="contained" key={`${assignment.id}_notdone`}>
```
```suggestion
          <MenuItem
            onClick={() => updateAssignmentStatus("not observed")}
            variant="contained"
            key={`${assignment.id}_notdone`}
          >
```

dannygoldstein replied:
Fixed

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+}
+        {
+          (isdone || isnotdone) && (
+          <MenuItem onClick={pendingAction} variant="contained" key={`${assignment.id}_pending`}>
```
```suggestion
          <MenuItem
            onClick={() => updateAssignmentStatus("pending")}
            variant="contained"
            key={`${assignment.id}_pending`}
          >
```

dannygoldstein replied:
Fixed

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+          )
+}
+        {isdone && (
+        <MenuItem key={`${assignment.id}_upload_spec`} onClick={handleClose}>
```
Can you help me understand what's supposed to happen here? The menu item says "Upload Spectrum" but its onClick is just tied to handleClose. Perhaps there's supposed to be a `Link` here as below with Upload Phot

dannygoldstein replied:
This is just a placeholder until the upload spectrum page is done.

dannygoldstein replied:
I can take this out if you think it shouldn't be integrated yet. Or I can comment it out.

acrellin replied:
How about just change the display text to include "Coming soon" or something, so it's clear it's not supposed to work yet?

dannygoldstein replied:
Done

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+            </Link>
+          </MenuItem>
+          )
+}
```
```suggestion
        }
```
soon, with forced `prettier`, we won't have to worry about these

dannygoldstein replied:
fixed

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+  const groups = useSelector((state) => state.groups.all);
+
+
+  // Load the observing run and its assignments if needed
```
This comment makes it sound like we're checking whether it's already loaded and only fetching if not, which we're not doing. Should that check be in the hook below?

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+        Loading run...
+      </b>
+    );
+  } else {
```
Please omit this else

dannygoldstein replied:
done

acrellin commented on static/js/components/RunSummary.jsx:
```diff
@@ -0,0 +1,423 @@
+      );
+    });
+
+
```
```suggestion
```

dannygoldstein replied:
fixed

stefanv commented on static/js/components/ThumbnailList.jsx:
```diff
@@ -3,31 +3,68 @@ import PropTypes from 'prop-types';
+
+const useStyles = makeStyles(() => ({
+  root: {
+    width: 200,
```
No absolute pixel sizes please.

dannygoldstein replied:
Fixed to use `rem` units

acrellin commented on static/js/components/StarList.jsx:
```diff
@@ -21,25 +38,49 @@ const StarList = ({ sourceId }) => {
+    );
+
+    const starlistInfo = [];
+    await Promise.allSettled(promises)
```
I don't think this is quite right; I think you should be using either `await` or `promise.then()` here, but not both. 
```suggestion
    values = await Promise.allSettled(promises);
    values.forEach(...);
```

dannygoldstein replied:
fixed

acrellin commented on static/js/components/ThumbnailList.jsx:
```diff
@@ -42,69 +79,76 @@ const Thumbnail = ({ ra, dec, telescope, mjd, name, url }) => {
-      ))}
-    </div>
-  );
+  if (useGrid) {
```
Are these changes supposed to be in here? Are they needed for the observing planner component?

dannygoldstein replied:
Yeah.

stefanv commented on static/js/components/VegaPlot.jsx:
```diff
@@ -166,20 +166,36 @@ const spec = (url) => ({
-);
+class VegaPlot extends React.Component {
+  // This is implemented as a class so we can define
+  // shouldComponentUpdate
```
Can we use hooks instead?

https://stackoverflow.com/a/56297625/214686

dannygoldstein replied:
Done.

stefanv commented on static/js/components/ThumbnailList.jsx:
```diff
@@ -3,31 +3,68 @@ import PropTypes from 'prop-types';
+    position: "absolute",
+    top: 0,
+    left: 0,
+    width: 200,
```
This is correct and needs to remain absolute.

stefanv commented on static/js/ducks/observingRun.js:
```diff
@@ -0,0 +1,69 @@
+messageHandler.add((actionType, payload, dispatch, getState) => {
+  const { observingRun } = getState();
+
+  /* TODO: Normalize state shape to eliminate the action processor below.
```
Please file a corresponding tracking issue, if it does not already exist.

dannygoldstein replied:
This is ch1046

acrellin commented on static/js/ducks/observingRun.js:
```diff
@@ -0,0 +1,69 @@
+  could be avoided in favor of a single update to the Source entity on the frontend.
+   */
+
+  if (actionType === REFRESH_SOURCE) {
```
Why not "REFRESH_ASSIGNMENT"? You're not re-fetching the whole source here

dannygoldstein replied:
Because the changes made to a source from the assignment page should also propagate to the source page, and vice versa. So to keep the `SourceHandler.put` method from getting saturated with many different types of websocket pushes, I just handle the "REFRESH_SOURCE" method within the `observingRun.js` reducer.

acrellin replied:
@dannygoldstein I think this would better be handled by a more specific WS message. The reason the followup request handler uses FETCH_SOURCE is because followup request data are fetched as part of the source data in GET /api/source/<source_id>, whereas assignments are fetched via their own unique endpoint. Just check whether the assignment was updated in SourceHandler.put and if so push a more specific message like REFRESH_ASSIGNMENT or something from there. In this case it's easy to make it more specific, and simpler.

dannygoldstein replied:
Will update the assignment reducer to look for REFRESH_ASSIGNMENT. But will have it push out REFRESH_SOURCE as well, as on the source page the logic is identical to that used for follow-up requests.

dannygoldstein replied:
Implemented REFRESH_OBSERVING_RUN

dannygoldstein commented on skyportal/handlers/api/internal/plot.py:
```diff
@@ -31,4 +36,27 @@ def get(self, obj_id):
+        obj = assignment.obj
+        permission_check = Source.get_obj_if_owned_by(obj.id, self.current_user)
+        if permission_check is None:
+            self.error('Invalid assignment id.')
```
Changed to "could not retrieve assignment."

acrellin replied:
Doesn't look like this change was pushed - not important here, but just want to make sure everything that should be pushed actually is.

dannygoldstein commented:
@stefanv @acrellin  Locally green and ready for re-review.

acrellin commented on skyportal/handlers/api/followup_request.py:
```diff
@@ -3,7 +3,7 @@
 from baselayer.app.access import auth_or_token
 from ..base import BaseHandler
-from ...models import (
+from ...models import (  # noqa
```
Isn't this already in master?

dannygoldstein replied:
Not according to github

acrellin replied:
What's being imported here that's not being used?

acrellin replied:
We should just remove the offending import rather than sweep it under the rug with the noqa comment. We use that sparingly in models.py for instance where we import things that aren't explicitly used there because other modules need to import them from models.py

acrellin commented on skyportal/tests/frontend/test_observing_planner.py:
```diff
@@ -0,0 +1,70 @@
+
+    submit_button = driver.wait_for_xpath('//*[@name="assignmentSubmitButton"]')
+
+    driver.scroll_to_element_and_click(submit_button)
```
We should check that the comment text appears in the assignment list after we click submit

dannygoldstein replied:
Fixed

acrellin commented:
Closing in favor of #711
